src/Data/LargeHashable/Intern.hs

```diff
@@ -6,67 +6,145 @@
+import Control.Monad
 import Data.Word
-import Foreign.Ptr
 import Foreign.C.Types
+import Foreign.Ptr
 import System.IO.Unsafe (unsafePerformIO)
-import qualified Data.ByteString as BS
-import qualified Data.ByteString.Char8 as BSC
-import qualified Data.ByteString.Base16 as Base16
-
--- | A newtype-wrapper for hash values.
-newtype Hash = Hash { unHash :: BS.ByteString }
-    deriving (Eq)
-
-instance Show Hash where
-    show (Hash bs) =
-        BSC.unpack (Base16.encode bs)
 
 -- | Functions for updating an intermediate hash value. The functions live
 -- in the 'IO' monad because they are typically implemented via FFI.
--- FIXME: functions for primitive types other than CULong missing.
 data HashUpdates
     = HashUpdates
-    { hu_updatePtr :: !(Ptr Word8 -> Int -> IO ()) -- ^ adds a byte array to the hash
-    , hu_updateULong :: !(CULong -> IO ())         -- ^ adds an `CULong` value to the hash
+    { hu_updatePtr :: {-# NOUNPACK #-} !(Ptr Word8 -> Int -> IO ()) -- ^ adds a byte array to the hash
+    , hu_updateChar :: {-# NOUNPACK #-} !(CChar -> IO ())      -- Int8
+    , hu_updateUChar :: {-# NOUNPACK #-} !(CUChar -> IO ())    -- Word8
+    , hu_updateShort :: {-# NOUNPACK #-} !(CShort -> IO ())    -- Int16
+    , hu_updateUShort :: {-# NOUNPACK #-} !(CUShort -> IO ())  -- Word16
+    , hu_updateInt :: {-# NOUNPACK #-} !(CInt -> IO ())        -- Int32
+    , hu_updateUInt :: {-# NOUNPACK #-} !(CUInt -> IO ())      -- Word32
+    , hu_updateLong :: {-# NOUNPACK #-} !(CLong -> IO ())      -- Int64
+    , hu_updateULong :: {-# NOUNPACK #-} !(CULong -> IO ())    -- Word64
     }
 
 -- | The interface for a hashing algorithm. The interface contains a simple run
 -- function, which is used to update the hash with all values needed, and the
 -- outputs the resulting hash.
-data HashAlgorithm
+data HashAlgorithm h
     = HashAlgorithm
-    { ha_run :: (HashUpdates -> IO ()) -> IO Hash
+    { ha_run :: {-# NOUNPACK #-} !((HashUpdates -> IO ()) -> IO h)
+    , ha_xor :: {-# NOUNPACK #-} !(h -> h -> h)
+    , ha_updateHash :: {-# NOUNPACK #-} !(HashUpdates -> h -> IO ())
+    }
+
+data LHEnv
+    = LHEnv
+    { lh_updates :: {-# NOUNPACK #-} !HashUpdates
+    , lh_updateXorHash :: {-# NOUNPACK #-} !([LH ()] -> IO ())
     }
 
 -- | The 'LH' monad (`LH` stands for "large hash") is used in the definition of
 -- hashing functions for arbitrary data types.
-newtype LH a = LH { _unLH :: ReaderT HashUpdates IO a }
-    deriving (Functor, Applicative, Monad)
+newtype LH a = LH (LHEnv -> IO a)
+
+{-# INLINE lhFmap #-}
+lhFmap :: (a -> b) -> LH a -> LH b
+lhFmap f (LH x) =
+    LH $ \env ->
+    do y <- x env
+       return (f y)
+
+{-# INLINE lhReturn #-}
+lhReturn :: a -> LH a
+lhReturn x = LH $ \_env -> return x
+
+{-# INLINE lhApp #-}
+lhApp :: LH (a -> b) -> LH a -> LH b
+lhApp (LH f) (LH x) =
+    LH $ \env -> f env <*> x env
 
+{-# INLINE lhBind #-}
+lhBind :: LH a -> (a -> LH b) -> LH b
+lhBind (LH x) f =
+    LH $ \env ->
+    do y <- x env
```

Comment:
shouldn't the bind be strict to avoid space leaks? Otherwise this looks like a lazy writer.